compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -30,6 +30,116 @@
 #include "il/Node_inlines.hpp"
 #include "infra/Bit.hpp"
 
+typedef enum RegisterStatus {
+   OK_FOR_REUSE,
+   DO_NOT_TOUCH
+} RegisterStatus;
+
+static inline RegisterStatus
+evaluateRegister(TR::Register **childReg, TR::Node *child, TR::Node *parent, TR::CodeGenerator *cg)
+   {
+   int64_t value;
+   RegisterStatus status;
+
+	if (child->getOpCode().isLoadConst() && child->getRegister() == NULL)
+   {
+      value = child->getLongInt();
+      if (constantIsUnsignedImm12(value))
+      {
+         //TODO: Cover special cases for immediate loads
+      }
+      else
+      {
+         *childReg = cg->allocateRegister();
+         loadConstant64(cg, parent, value, *childReg);
+         status = OK_FOR_REUSE;
```

Comment:
The register re-use mechanism should be driven by the reference count on the node whose value the register represents.  When you evaluate a node, its reference count represents the number of remaining uses of that node in subsequent trees (including the current one).  You can derive some properties of the register you need based on the reference count on the node.  For instance, if the reference count on the node you are evaluating is 1 then you know that this is the last unseen reference to the node in the trees and the register that represents its result effectively goes dead after this point (i.e., the register can be reused).  Furthermore, if the reference count on a register is >=1 and there hasn't been a register yet associated with a node then that tells you that this is the first occurrence of that node in the trees and hence this is the first evaluation of it.  A node with a reference count >1 tells you that the live range of the register associated with it does not die at this point.

I think your register status mechanism should be driven by the reference count and the register status on the node.  Perhaps you can think about whether the status mechanism is really required (and if so, why is it needed over and above the reference counting mechanism).

Reply:
Thank you for the illustration, it is very helpful! We'll look into changing this as well.